### PR TITLE
[Perf] Enable torch.compile on MammothModa2 DiT Stage

### LINKED ISSUE #7139: [Perf] MammothModa2 DiT: enable and validate regional torch.compile
## Summary

Parent issue: #7075
Owner: @Xenoryn

Enable regional torch.compile for the MammothModa2 DiT transformer, starting with Preview text-to-image in single-request/request-mode execution. Report Dev text-to-image validation separately.

## Scope

- Identify suitable transformer compilation regions and address graph breaks within those regions.
- Ensure compilation configuration reaches the executing backend.
- Validate supported resolution, conditioning sequence-length, and CFG shape variations; document specialization and recompilation behavior.
- Preserve conditioning, masks, RoPE, dtype boundaries, and denoising semantics.
- Document supported configurations and explicit errors or fallbacks for unsupported cases.

Shared-runtime migration, CUDA graph capture/replay, VAE compilation, caching, quantization, offload, batching, and parallelism implementation are outside the initial scope.

## Dependencies and interfaces

- Coordinate with @Levius-Fubuki (#7086) on the shared-runtime interface, model initialization, and compilation setup. The initial integration target remains request mode with max_num_seqs=1.
- Coordinate with @xsmccc on compilation/warmup versus CUDA graph capture boundaries and subsequent combined validation.
- Align transformer module and metadata interfaces with @Jerry2423, @MrlixiangWE, and @TonybotNi.
- Coordinate CFG shape expectations with @ShengleiFu without making packed CFG implementation a prerequisite for the initial scope.

Model-level compilation experiments can proceed while runtime integration is being finalized; E2E support requires validation on the agreed runtime.

## Planned PRs and schedule

1. Regional compilation integration and required compatibility fixes.
2. Shape coverage, E2E validation, and documentation, as a follow-up PR if needed.

Proposed PR-ready date: 2026-09-12 (UTC).
Proposed target merge date: 2026-09-17 (UTC).
Dates are tentative and depend on #7086 interface readiness and GPU access.

PR links: to be added. Partial PRs will use `Refs #7075`.
Reviewer: maintainer review requested.

## Validation and performance evidence

Hardware/backend: GPU model, GPU count, backend, and availability to be confirmed.

Use identical eager/compiled configurations with pinned model/code revisions, dtype, prompts, seeds, and device placement. Start with 1024×1024, 50 denoising steps, guidance 4.0, and seed 42, then extend to additional supported resolutions and conditioning lengths.

- Compare eager/compiled transformer outputs and denoising latents using identical conditioning and initial noise, with documented tolerances.
- Validate complete image generation and image-quality parity.
- Exercise repeated requests and shape changes, recording graph breaks and recompilation behavior.
- Confirm compiled regions execute rather than silently falling back.
- Measure compilation/warmup separately from steady-state transformer, DiT-stage, and E2E latency, plus peak memory.
- Isolate compilation initially, then validate supported CUDA graph combinations jointly with their owner.

Publish reproducible commands and paired baseline/head evidence. Untested variants/backends/combinations will be explicitly identified.

## Blockers and availability

Runtime integration depends on agreement with #7086.
Compilation blockers and supported shape coverage remain to be audited.
Hardware availability and personal schedule: to be confirmed.

### PR DESCRIPTION
**Pull request by Xenoryn**
## Purpose

Enable `torch.compile` on the MammothModa2 DiT stage (`Transformer2DModel`) with `fullgraph=True`, no graph break, no shape-driven recompile across resolutions and CFG shapes.

Sub-issue of #7075: closes #7139.

**Depends on #7134** (@Levius-Fubuki's runtime migration; sub-issue #7086). Landing plan below.

AR compile / DiT CUDA graph capture / TeaCache / TP / SP are tracked in their own PRs and out of scope here.

## Landing plan

The commit history of this branch is intentionally two commits:

1. `[Model][MammothModa2] Enable regional torch.compile on DiT TransformerBlock` — runner-agnostic core.
2. `[TEMPORARY][MammothModa2] Bridge regionally_compile from model __init__` — 17-line hook that will be **dropped on rebase after #7134 lands**.

⚠️ Once #7134 flips stage 1 to `StageExecutionType.DIFFUSION`, `DiffusionModelRunner._compile_transformer` reads `pipeline._dit_modules` and drives `regionally_compile` for us; the second commit here becomes redundant. This PR stays in **Draft** until #7134 merges; we then rebase, drop commit 2, re-run the benchmarks below on the post-migration baseline, and mark ready for review.

## What changed in the DiT forward

Block topology is untouched. Three small edits let the block body trace clean under `fullgraph=True`, plus one class attribute wires the compile target.

**Declare the compile target.** `Transformer2DModel` gets `_repeated_blocks = ["TransformerBlock"]`. Without it `regionally_compile` doesn't know what to walk into and no-ops silently — which is what was happening before this PR.

**Lift the empty-sequence skip out of the compiled block.** Under CFG the unconditional pass hands the context refiners a zero-token text stream, and the old block-level `if sequence_length == 0` short-circuit is a data-dependent branch on a SymInt — Dynamo won't trace it under `fullgraph`. Moved the same skip up to `_apply_refiners`, which runs in eager Python where `shape[1]` is a real int. On an empty text stream `context_refiner` collapses to identity anyway (no modulation, residual = input), so this is a semantic no-op.

**Stop tracing a `logger.debug` call.** `sdpa.py`'s GQA-expansion path had a debug log inside the traced region; Dynamo rejects `logging.Logger` calls under `fullgraph`. Guarded it with `torch.compiler.is_compiling()`.

**Temporary hook to actually trigger the compile.** Stage 1 is currently `LLM_GENERATION`, whose runner never calls `_compile_transformer`. Added a 17-line block in `MammothModa2ForConditionalGeneration.__init__` that calls `regionally_compile` directly when `enforce_eager=False`. Gets removed on rebase after #7134.

## How to enable

Stage 1 (DiT) requires `enforce_eager: false` in the deploy YAML, plus two workarounds that only exist because we're still on the LLM_GENERATION path:

- `cudagraph_mode: NONE` — MammothModa2's DiT pipeline has an unpinned CPU→CUDA copy on `text_cond` / `image_cond` (`pipeline_mammothmoda2_dit.py:260`, `.to(non_blocking=True)` without `pin_memory`) that breaks vLLM's CUDA graph capture during warmup. Cudagraph is orthogonal to what this PR measures.
- `custom_ops: [all]` — when `enforce_eager: false` on an LLM_GENERATION stage, vLLM 0.29 flips this stage's default from `['all']` to `['none']` and drops `vllm_c` from `rms_norm`'s IR priority, forcing every RMSNorm outside the compiled region (pipeline embeddings, scheduler step, VAE decode) onto native PyTorch. Adds ~2 s to E2E.

Both go away after #7134: `DiffusionModelRunner` does not route through vLLM's `CompilationConfig`.

Reference deploy YAML used for the benchmarks below: `tests/diffusion/models/mammoth_moda2/bench_t2i_compile.yaml` (kept in the branch for reproducibility of the numbers in this PR; will be removed on the final rebase).

## Test Plan

- **Compile parity unit tests**: `pytest tests/diffusion/models/mammoth_moda2/test_dit_compile_parity.py` — drives the same code path the runner uses, `regionally_compile(model, dynamic=True, fullgraph=True)`, and asserts `torch.equal(eager, compiled)` across `T ∈ {0, 4, 77}` (the T=0 case exercises the moved empty-seq guard), two resolutions, and a CFG-style `B=2` batch. Also probes `torch._dynamo.utils.counters` to assert no per-block recompile between shapes.
- **Empty-sequence coverage at the attention layer**: existing `tests/diffusion/models/mammoth_moda2/test_dit_attention.py` and `test_dit_attention_cuda.py` still cover the `AttnProcessor` empty-input case, just no longer via the removed short-circuit — the guard is now enforced by `_apply_refiners`.
- **Real-model end-to-end** on H800 with `MammothModa2-Dev` — four configs side by side: `main-eager` / `branch-eager` / `branch-compile`.

**vLLM Version:** 0.29.0
**vLLM-Omni Commit:** based on `bc0c9f4b`

## Test Result

**Compile parity** — all cases pass under `fullgraph=True, dynamic=True` with `torch.equal` (bitwise). Recompile probe reports zero additional compilations across the shape sweep.

**No graph break / no recompile** — `TORCH_LOGS="graph_breaks,recompiles"` on the E2E run below emits zero entries across DiT warmup + 50 steps × 2 CFG batches.

### End-to-end T2I benchmark

Nvidia H800. `MammothModa2-Dev`, 1024×1024, 50 DiT steps, guidance 4.0, CFG range `[0.0, 1.0]`, seed 42. Prompt: *"A stylish woman riding a motorcycle in NYC, movie poster style"*. Cadence per config: 1 warmup (4 DiT steps) + 3 timed runs.

Cross-process DiT timing via CUDA events injected into the stage-1 engine core subprocess (`sitecustomize.py` + `bench_probe.py`); E2E is wall-clock around `omni.generate`. Attention runs on **FA3** for both stages.

**Cumulative DiT-forward time** (50 steps × 2 CFG batches = 100 forwards per run):

| Config | DiT mean ± σ (ms) | vs main-eager | vs branch-eager |
| --- | ---: | ---: | ---: |
| main-eager | 9,133 ± 14 | — | +36.7% |
| branch-eager | 6,678 ± 12 | **1.37×** faster | — |
| branch-compile | **3,809 ± 6** | **2.40×** faster | **1.75×** faster |

**End-to-end wall-clock**:

| Config | E2E mean ± σ (ms) | E2E − DiT (ms) | vs main-eager E2E |
| --- | ---: | ---: | ---: |
| main-eager | 106,443 ± 19 | 97,310 | — |
| branch-eager | 103,670 ± 515 | 96,992 | 1.027× |
| **branch-compile** | **100,939 ± 223** | 97,128 ✓ | **1.055×** |

**On the E2E ratio**: the AR stage dominates the wall-clock (≈97 s of the 100–106 s total, generating ~1,057 visual tokens); DiT is only 3.6–8.6% of E2E, so a 2.40× DiT speedup translates to a modest 1.055× E2E speedup by Amdahl. AR compile is being tracked separately (#7075 → @Xenoryn), and lands the bulk of the remaining E2E headroom.

**branch-eager is 27% faster on DiT than main-eager** — this is the unintended side-effect of moving the empty-seq guard: on main, the CFG unconditional pass launches `attn.to_out[1](attn.to_out[0](empty))` per block per step even when there's nothing to attend to (32 blocks × 50 steps × 2 CFG batches = 3,200 empty-kernel launches per image). The new outer-level skip removes them all.

### Image outputs

TODO

### CODE REVIEW

File: tests/diffusion/models/mammoth_moda2/test_dit_compile_parity.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    # Strict fp32 tolerance rather than torch.equal: Inductor's kernel fusion
+    # legitimately reorders FMAs. The bitwise contract from #7139 applies to
+    # the AR→DiT hidden-state boundary — which the compile change does not
+    # touch (same tensor object flows through ``Transformer2DModel.forward``).
+    assert torch.allclose(want, got, rtol=1e-5, atol=1e-5), (
+        f"max abs diff: {(want - got).abs().max().item()}"
+    )
+
+
+def test_forward_no_shape_driven_recompile_across_resolutions_and_text_len():
```
Comment (MrlixiangWE) [inline tests/diffusion/models/mammoth_moda2/test_dit_compile_parity.py:141]: Run with the rest of this file, this test fails with `regional compile did not trigger on first shape` (`assert 0 >= 1`; 1 failed, 3 passed) and passes on its own. The parity cases above compile the same `TransformerBlock.forward` code object with `dynamic=True`, so the first shape here hits dynamo's cache and the counter sees no compile. The file is `core_model` and `cpu`, which puts it in the ready and merge "Simple · Diffusion Test" lane that runs `tests/diffusion` in one process. With `torch._dynamo.reset()` as the first line of this test, the file passes (4 passed), and so do `tests/diffusion/models/mammoth_moda2` and `tests/diffusion/layers/test_norm.py` (99 passed).

**Review by MrlixiangWE — COMMENTED** [review]
Measured with this PR merged onto main (`cb5f508`) on 1x A800 with vLLM 0.30.0, MammothModa2-Preview text-to-image at 1024x1024, 50 steps, guidance 4.0, seed 42, studio-cat and Samoyed prompts, FLASH_ATTN, mean of 4 timed requests (2 per prompt after one warmup), the two runs back to back on one GPU:

| Stage 1 | DiT stage time |
| --- | ---: |
| `enforce_eager: true` | 15,640 ms |
| `enforce_eager: false` (regional compile) | 10,968 ms (-29.87%) |

Compiled images against eager (PSNR 50.05 dB for the cat, 25.16 dB for the Samoyed):

Eager and compiled DiT, studio cat and Samoyed, seed 42

Compiled, it also runs with Cache-DiT, TORCH_SDPA, #7969 and #7489. On main, `enforce_eager: false` only logs "Regional compilation skipped" and the images match eager bit for bit.

On a compiled FLASH_ATTN run of the cat prompt with `TORCH_LOGS=graph_breaks,recompiles` (line numbers at main `cb5f508`), the compiled blocks break at `vllm_omni/diffusion/attention/backends/flash_attn.py:463` (`if attention_mask is not None and torch.any(~attention_mask)`, data-dependent branching; the same branch is at line 450 of this PR's base). On this model's path the check sits in front of the kernel call whichever FA version is loaded. With only vLLM's bundled FA2 available, `flash_attn_func` is None, so each block breaks again in the dense fallback (`flash_attn.py:483`), at `_vllm_fa2_C.varlen_fwd`, which has no fake-tensor kernel. The two block kinds (`norm1` is `Qwen2RMSNorm` in the context refiner and `LuminaRMSNormZero` elsewhere) compile once each during startup, which is expected.

Stage 1 in `vllm_omni/deploy/mammoth_moda2.yaml` keeps `enforce_eager: true`, and the recipe does not mention compile; a commented `enforce_eager: false` next to the Cache-DiT block would show how to turn it on.
